src/App.tsx

```diff
@@ -15,36 +17,109 @@ export const goodsFromServer = [
+  alphabetically = 'alphabetically',
+  byLength = 'byLength',
+  default = '',
+}
+
+type SortObj = {
+  sortField: SortType;
+  reverseGood?: boolean;
+};
+
+function getPreparedGoods(
+  goods: string[],
+  { sortField, reverseGood = false }: SortObj) {
+  const preparedGoods = [...goods];
+
+  if (sortField) {
+    preparedGoods.sort((good1, good2) => {
+      switch (sortField) {
+        case SortType.alphabetically:
+          return good1.localeCompare(good2);
+        case SortType.byLength:
+          return good1.length - good2.length;
+        default:
+          return 0;
+      }
+    });
+  }
+
+  if (reverseGood) {
+    preparedGoods.reverse();
+  }
+
+  return preparedGoods;
+}
+
 export const App: React.FC = () => {
+  const [sortField, setSortField] = useState<SortType>(SortType.default);
+  const [reverseGood, setReverse] = useState(false);
+  const visibleGoods = getPreparedGoods(goodsFromServer, {
+    sortField,
+    reverseGood,
+  });
+
+  const reset = () => {
+    setSortField(SortType.default);
+    setReverse(false);
+  };
+
   return (
     <div className="section content">
       <div className="buttons">
-        <button type="button" className="button is-info is-light">
+        <button
+          type="button"
+          className={cn('button is-info', {
+            'is-light': sortField !== SortType.alphabetically,
+          })}
+          onClick={() => {
+            setSortField(SortType.alphabetically);
+          }}
+        >
           Sort alphabetically
         </button>
-
-        <button type="button" className="button is-success is-light">
+        <button
+          type="button"
+          className={cn('button is-success', {
+            'is-light': sortField !== SortType.byLength,
+          })}
+          onClick={() => {
+            setSortField(SortType.byLength);
+          }}
+        >
           Sort by length
         </button>
-
-        <button type="button" className="button is-warning is-light">
+        <button
+          type="button"
+          className={cn('button is-warning', {
+            'is-light': !reverseGood,
+          })}
+          onClick={() => {
+            setReverse(!reverseGood);
+          }}
+        >
           Reverse
         </button>
 
-        <button type="button" className="button is-danger is-light">
-          Reset
-        </button>
+        {(sortField || reverseGood) && (
+          <button
+            type="button"
+            className="button is-danger is-light"
+            onClick={reset}
+          >
+            Reset
+          </button>
+        )}
       </div>
 
       <ul>
-        <ul>
-          <li data-cy="Good">Dumplings</li>
-          <li data-cy="Good">Carrot</li>
-          <li data-cy="Good">Eggs</li>
-          <li data-cy="Good">Ice cream</li>
-          <li data-cy="Good">Apple</li>
-          <li data-cy="Good">...</li>
-        </ul>
+        {visibleGoods.map(good => (
+          <li data-cy="Good" key={good}>
```

Comment:
The `key` prop for list items should be a unique identifier. Using `good` as a key is risky if the list can contain duplicate values. To comply with the [CODE KNOWLEDGE] checklist, ensure that the `key` is unique, possibly by combining the item with its index or using another unique value.